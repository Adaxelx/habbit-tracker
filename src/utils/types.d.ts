export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: boolean;
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}
