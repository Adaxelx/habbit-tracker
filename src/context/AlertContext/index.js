export { default, useAlertContext } from './AlertContext';
