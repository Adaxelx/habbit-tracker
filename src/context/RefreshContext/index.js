export { default, useRefreshContext } from './RefreshContext';
