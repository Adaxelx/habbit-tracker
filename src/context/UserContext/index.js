export { default, useUserContext, UserContext } from './UserContext';
