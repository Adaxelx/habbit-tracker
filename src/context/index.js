export { default as ThemeProvider, useThemeContext } from './ThemeContext';
export { default as UserProvider, useUserContext, UserContext } from './UserContext';
export { default as AlertProvider, useAlertContext, AlertContext } from './AlertContext';
export { default as RefreshProvider, useRefreshContext } from './RefreshContext';
