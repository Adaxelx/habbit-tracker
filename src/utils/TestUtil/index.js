export { default } from './TestUtil.ts';
