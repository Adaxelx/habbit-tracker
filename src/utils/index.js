export { default as TestUtil } from './TestUtil';
export { default as RenderWithRouter } from './RenderWithRouter';
export * from './validators';
export * from './types.d';
