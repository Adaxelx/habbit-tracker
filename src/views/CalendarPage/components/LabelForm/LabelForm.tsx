import React, { useRef } from 'react';
import { PopUp, Button, Input } from 'components';
import { useForm, Controller } from 'react-hook-form';
import { useAlertContext, useUserContext } from 'context';
import { FormProps, LabelSend, createRestrictedLengthObject, AlertTypes } from 'utils';
import { SketchPicker } from 'react-color';
import { StyledWrapper } from 'views/CalendarPage/components/HabbitForm/HabbitForm.css';
import { postLabel } from 'views/CalendarPage/CalendarPage.api';

const { SUCCESS } = AlertTypes;

const LabelForm = ({ handleClose, open, handleRefresh }: FormProps) => {
  const { register, handleSubmit, errors, control } = useForm<LabelSend>();

  const alertC = useRef(useAlertContext());
  const { token } = useUserContext();

  const onSubmit = async (data: LabelSend) => {
    try {
      await postLabel(token, data);
      alertC.current.showAlert('Succesfuly added label to your calendar.', SUCCESS);
      handleClose();
      handleRefresh();
    } catch (err) {
      alertC.current.showAlert(err.message);
    }
  };

  return (
    <PopUp open={open} handleClose={handleClose} header="Add label">
      <StyledWrapper as="form" onSubmit={handleSubmit(onSubmit)} noValidate>
        <Input
          name="title"
          label="Title"
          id="title"
          refVal={register(createRestrictedLengthObject('title'))}
          minLength={6}
          maxLength={30}
          error={errors.title}
          data-testid="title"
        />
        <Controller
          name="color"
          control={control}
          render={({ onChange, value }) => (
            <SketchPicker color={value} onChangeComplete={(color) => onChange(color.hex)} />
          )}
        />
        <Button size="s" mt="16px" type="submit" data-testid="submit" noMaxWidth>
          Send
        </Button>
      </StyledWrapper>
    </PopUp>
  );
};

export default LabelForm;
