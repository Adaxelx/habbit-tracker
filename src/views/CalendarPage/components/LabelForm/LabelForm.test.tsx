import React from 'react';
import 'jest-styled-components';
import { waitFor } from '@testing-library/dom';
import { AlertContext } from 'context';
import { AlertTypes, FormLabel, TestUtil } from 'utils';
import LabelForm from './LabelForm';

jest.mock('react-color', () => ({
  SketchPicker: ({ onChangeComplete, value }) => (
    <input data-testid="color" onChange={onChangeComplete} value={value} />
  ),
}));

describe('LabelForm', () => {
  let util: TestUtil;
  let props: FormLabel;
  let fail: boolean;
  const alertC = { showAlert: jest.fn() };

  const mockedFetch = () => {
    if (fail) {
      return Promise.resolve({
        status: 500,
        ok: false,
        json: () => Promise.resolve({ message: 'error' }),
      });
    }

    return Promise.resolve({
      status: 201,
      ok: true,
    });
  };

  beforeAll(() => {
    jest.spyOn(window, 'fetch');
  });

  beforeEach(() => {
    props = {
      handleClose: jest.fn(),
      open: true,
    };
    fail = false;
    jest.clearAllMocks();
    window.fetch.mockImplementation(mockedFetch);
  });

  it('should match snapshot', () => {
    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <LabelForm {...props} />
        </AlertContext.Provider>
      ),
    );
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('should send label', async () => {
    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <LabelForm {...props} />
        </AlertContext.Provider>
      ),
    );

    util.setValue('title', 'title123');
    util.setValue('color', '#ffffff');
    util.click('submit');
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('Succesfuly added label.', AlertTypes.SUCCESS);
  });

  it('should not send Habit if api fails', async () => {
    fail = true;

    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <LabelForm {...props} />
        </AlertContext.Provider>
      ),
    );

    util.setValue('title', 'title123');
    util.setValue('color', '#ffffff');
    util.click('submit');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('error');
  });

  it('should send edited Habit', async () => {
    props = {
      ...props,
      label: { _id: '1', userId: '123', color: '#fffffff', title: 'label' },
    };

    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <LabelForm {...props} />
        </AlertContext.Provider>
      ),
    );

    util.setValue('title', 'title1234343');
    util.click('submit');

    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('Succesfuly edited label.', AlertTypes.SUCCESS);
  });
});
