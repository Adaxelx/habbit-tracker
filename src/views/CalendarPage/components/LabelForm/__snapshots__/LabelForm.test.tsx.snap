// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelForm should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
  z-index: 1000;
  background-color: rgba(0,0,0,0.3);
}

<div
    class="c0"
  />
  .c5 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
  width: 100%;
  max-width: 300px;
}

.c7 {
  font-size: 1.5rem;
  margin-bottom: 0.75rem;
}

.c8 {
  font-size: 1.25rem;
  width: 100%;
  max-width: 300px;
}

.c8:focus {
  outline: thin dotted;
}

.c9 {
  color: #fff;
  background-color: #FFC136;
  border: 1px solid #707070;
  padding: 0.75rem;
  font-size: 1rem;
  min-width: 150px;
  display: inline-block;
  text-align: center;
  margin-top: 16px;
  margin-left: 0;
  margin-right: 0;
}

.c9:hover {
  cursor: pointer;
}

.c9:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

.c0 {
  position: fixed;
  width: calc(100% - 1rem - 1rem);
  height: calc(90vh - 1rem - 1rem);
  top: 1rem;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  padding: 1rem 0;
  background-color: #AFE9FF;
  border: 1px solid #707070;
  z-index: 1000;
  max-width: 350px;
  overflow: scroll;
}

.c4 {
  width: 100%;
  text-align: center;
  font-size: 2rem;
}

.c1 {
  width: 50px;
  height: 50px;
  position: relative;
  display: block;
  margin-left: auto;
  margin-right: 0;
  border: none;
  background-color: transparent;
}

.c1:hover {
  cursor: pointer;
}

.c2 {
  height: 100%;
  width: 2px;
  position: absolute;
  top: 0%;
  left: 50%;
  background-color: #FF9191;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c3 {
  height: 100%;
  width: 2px;
  position: absolute;
  top: 0%;
  left: 50%;
  background-color: #FF9191;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="Add label-close"
      type="button"
    >
      <div
        class="c2"
        rotate="45"
      />
      <div
        class="c3"
        rotate="-45"
      />
    </button>
    <h2
      class="c4"
      data-testid="Add label-header"
    >
      Add label
    </h2>
    <form
      class="c5"
      novalidate=""
    >
      <div
        class="c6"
      >
        <label
          class="c7"
          for="title"
        >
          Title
        </label>
        <input
          class="c8"
          data-testid="title"
          id="title"
          maxlength="30"
          minlength="6"
          name="title"
          type="text"
        />
      </div>
      <input
        data-testid="color"
        value=""
      />
      <button
        class="c9"
        data-testid="submit"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</DocumentFragment>
`;
